node/builder_miner.go

```diff
@@ -0,0 +1,223 @@
+		Override(new(*stores.Remote), modules.RemoteStorage),
+		Override(new(dtypes.RetrievalPricingFunc), modules.RetrievalPricingFunc(cfg.Dealmaking)),
+
+		If(!cfg.Subsystems.EnableMining,
+			If(cfg.Subsystems.EnableSealing, Error(xerrors.Errorf("sealing can only be enabled on a mining node"))),
+			If(cfg.Subsystems.EnableSectorStorage, Error(xerrors.Errorf("sealing can only be enabled on a mining node"))),
+		),
+		If(cfg.Subsystems.EnableMining,
+			If(!cfg.Subsystems.EnableSealing, Error(xerrors.Errorf("sealing can't be disabled on a mining node yet"))),
+			If(!cfg.Subsystems.EnableSectorStorage, Error(xerrors.Errorf("sealing can't be disabled on a mining node yet"))),
+
+			// Sector storage: Proofs
+			Override(new(ffiwrapper.Verifier), ffiwrapper.ProofVerifier),
+			Override(new(ffiwrapper.Prover), ffiwrapper.ProofProver),
+			Override(new(storage2.Prover), From(new(sectorstorage.SectorManager))),
+
+			// Sealing (todo should be under EnableSealing, but storagefsm is currently bundled with storage.Miner)
+			Override(new(sealing.SectorIDCounter), modules.SectorIDCounter),
+			Override(GetParamsKey, modules.GetParams),
+
+			Override(new(dtypes.SetSealingConfigFunc), modules.NewSetSealConfigFunc),
+			Override(new(dtypes.GetSealingConfigFunc), modules.NewGetSealConfigFunc),
+
+			// Mining / proving
+			Override(new(*slashfilter.SlashFilter), modules.NewSlashFilter),
+			Override(new(*storage.Miner), modules.StorageMiner(config.DefaultStorageMiner().Fees)),
+			Override(new(*miner.Miner), modules.SetupBlockProducer),
+			Override(new(gen.WinningPoStProver), storage.NewWinningPoStProver),
+			Override(new(*storage.Miner), modules.StorageMiner(cfg.Fees)),
+			Override(new(sectorblocks.SectorBuilder), From(new(*storage.Miner))),
+		),
+
+		If(cfg.Subsystems.EnableSectorStorage,
+			// Sector storage
+			Override(new(*stores.Index), stores.NewIndex),
+			Override(new(stores.SectorIndex), From(new(*stores.Index))),
+			Override(new(*sectorstorage.Manager), modules.SectorStorage),
+			Override(new(sectorstorage.Unsealer), From(new(*sectorstorage.Manager))),
+			Override(new(sectorstorage.SectorManager), From(new(*sectorstorage.Manager))),
+			Override(new(storiface.WorkerReturn), From(new(sectorstorage.SectorManager))),
+		),
+
+		If(!cfg.Subsystems.EnableSectorStorage,
+			Override(new(sectorstorage.StorageAuth), modules.StorageAuthWithURL(cfg.Subsystems.SectorIndexApiInfo)),
+			Override(new(modules.MinerStorageService), modules.ConnectStorageService(cfg.Subsystems.SectorIndexApiInfo)),
+			Override(new(sectorstorage.Unsealer), From(new(modules.MinerStorageService))),
+			Override(new(sectorblocks.SectorBuilder), From(new(modules.MinerStorageService))),
+		),
+		If(!cfg.Subsystems.EnableSealing,
+			Override(new(modules.MinerSealingService), modules.ConnectSealingService(cfg.Subsystems.SealerApiInfo)),
+			Override(new(stores.SectorIndex), From(new(modules.MinerSealingService))),
+		),
+
+		If(cfg.Subsystems.EnableStorageMarket,
```

Comment:
Why don't we include the `LibP2P` option here (and in the client options)? This should allow us to get rid of `settings.enableLibp2pNode`

Reply:
I will give this a try, but I am not sure it is possible, as there are some configs that need to be applied earlier.

Reply:
Agree. Now that we've separated `Libp2p` from `Online`, `enableLibp2pNode` maps directly to enabling the `Libp2p` aggregate. This should be feasible.

Reply:
Oh wait, I think this may have to do with the fact that we need to read the Config to know whether to enable or not the libp2p node, right?

Reply:
Only when Markets is enabled in the miner config, should we enable libp2p. So it's a bit of a chicken and egg, and that's why it's convoluted.

Reply:
Actually, no. It should be possible!